text/2022-11-08-font-compiler-ir.md

```diff
@@ -0,0 +1,190 @@
+
+## Potential future goals
+
+Enable translation from `sourceA → IR → sourceB` so we can modernise sources in an automated manner (credit @cmyr for this idea).
+
+## Non-goals
+
+Lossless `source → IR → source`.
+
+# Basic structure
+
+```
+Source → IR → font binary
+            ^ one ir2font implementation
+       ^ many source2ir implementations
+```
+
+Source is one of
+
+- .glyphs
+- UFO (including many w/design space)
+- .ttx
+- nanoemoji config
+
+In future, sources could additionally include:
+
+- binary font
+  - Some contend if we support ttx we effectively support binary fonts
+- Unknown Future Design Format
+
+For every source we will have a dedicated conversion to IR, e.g. `glyphs2ir`, `ufo2ir`, etc in https://github.com/googlefonts/fontmake-rs.
+
+Let’s take `ufo2ir` as a concrete example, it would:
+
+- Depend on a UFO parser, likely https://github.com/linebender/norad
+  - Aside: norad uses quick-xml, we should try to stick to that elsewhere
+- Depend on the IR library
+
+# IR
+
+## Font sources
+
+Rough outline of common source structures
+
+UFO:
+
+- Designspace file maps UFO(s) into design space
+- UFO font has layers
+- Layers have glyphs
+
+Glyphs:
+
+- Font has glyphs
+- Glyphs have one layer per master
+- Plus user defined layers
+- Plus "special" layers that have attributes
+  - For example, "brace" layers to define intermediate masters for specific glyphs
+
+## Sketch
+
+Based on [ufo3](https://unifiedfontobject.org/versions/ufo3/), [babelfont-rs](https://github.com/simoncozens/babelfont-rs) and discussion in this repo. The
+structures sketched here in pseudo-Rust (based on multiple reports a similar format was helpful for colr). We deliberately have no concept of "master", we
+instead have a set of variable glyphs whose definition locations need not be identical. The examples are not exhaustive of all fields, they merely aspire to
+convey the general idea of a variable first, sparse first, structure broken apart in alignment with units of work for a parallel backend:
+
+```rust
+// data common to the entire designspace
+// written to disk in one file
+// some of https://unifiedfontobject.org/versions/ufo3/fontinfo.plist/ 
+struct Font
+  upem
+  axes: Vec<Axis>
+
+// TODO: do we really need "discrete" axes? - designspace 5.0
+enum AxisRange {
+  Continuous(min, max, default)
+  Discrete(Vec<f32>)  // e.g. Italic
+}
+struct Axis { tag, range: AxisRange, hidden }
+
+// Roughly
+type DesignSpaceLocation = HashMap<Tag, f32>
+
+// the complete variable definition of a single glyph
+// written to disk in one file per glyph
+// components point to other glyphs, in time expand to variable components
+// https://github.com/simoncozens/rust-font-tools/blob/main/babelfont-rs/src/glyph.rs 
+// an obvious unit of work for parallel execution
+// Note that this is sparse, not tied to masters, and the definition keys need not match glyph to glyph
+struct Glyph
+  name
+  definitions: Map<DesignSpaceLocation, GlyphInstance>
```

Comment:
FWIW, in Fontra we separate "layers" from "sources". A *layer* refers to a glyph instance, a *source* specifies the designspace location and refers to a layer. This extra indirection allows a *layer* to be used by multiple *sources*, which is a common pattern to tame designspaces that otherwise may have "gaps", and it is a necessary feature when implementing HOI-like behavior.

Our dataclasses-based schema so far: https://github.com/BlackFoundryCom/fontra/blob/main/src/fontra/core/classes.py

Reply:
I took the liberty of dumping Fontra's json schema below, in case it might be of inspiration for the proposed IR.

```js
$ python src/fontra/core/classes.py
// This file is generated, don't edit!
export const classSchema = {
  "VariableGlyph": {
    "name": {
      "type": "str"
    },
    "axes": {
      "type": "list",
      "subtype": "LocalAxis"
    },
    "unicodes": {
      "type": "list",
      "subtype": "int"
    },
    "sources": {
      "type": "list",
      "subtype": "Source"
    },
    "layers": {
      "type": "list",
      "subtype": "Layer"
    }
  },
  "LocalAxis": {
    "name": {
      "type": "str"
    },
    "minValue": {
      "type": "float"
    },
    "defaultValue": {
      "type": "float"
    },
    "maxValue": {
      "type": "float"
    }
  },
  "Source": {
    "name": {
      "type": "str"
    },
    "layerName": {
      "type": "str"
    },
    "location": {
      "type": "dict"
    }
  },
  "Layer": {
    "name": {
      "type": "str"
    },
    "glyph": {
      "type": "StaticGlyph"
    }
  },
  "StaticGlyph": {
    "path": {
      "type": "PackedPath"
    },
    "components": {
      "type": "list",
      "subtype": "Component"
    },
    "xAdvance": {
      "type": "float",
      "optional": true
    },
    "yAdvance": {
      "type": "float",
      "optional": true
    },
    "verticalOrigin": {
      "type": "float",
      "optional": true
    }
  },
  "PackedPath": {
    "coordinates": {
      "type": "list",
      "subtype": "float"
    },
    "pointTypes": {
      "type": "list",
      "subtype": "PointType"
    },
    "contourInfo": {
      "type": "list",
      "subtype": "ContourInfo"
    }
  },
  "ContourInfo": {
    "endPoint": {
      "type": "int"
    },
    "isClosed": {
      "type": "bool"
    }
  },
  "Component": {
    "name": {
      "type": "str"
    },
    "transformation": {
      "type": "Transformation"
    },
    "location": {
      "type": "dict"
    }
  },
  "Transformation": {
    "translateX": {
      "type": "float"
    },
    "translateY": {
      "type": "float"
    },
    "rotation": {
      "type": "float"
    },
    "scaleX": {
      "type": "float"
    },
    "scaleY": {
      "type": "float"
    },
    "skewX": {
      "type": "float"
    },
    "skewY": {
      "type": "float"
    },
    "tCenterX": {
      "type": "float"
    },
    "tCenterY": {
      "type": "float"
    }
  }
}
```

Reply:
- Note that none of this is set in stone in Fontra. 
- `PackedPath` should be mostly seen as an internal datastructure, which I've optimized for network transmission and interpolation, but is too low-level as a source format. It is modeled after the TTF glyf data structure. Perhaps it is not even that bad for an intermediate respresentation, but I wouldn't immediately recommend it
- I'm still torn about `glyph.unicodes` vs a cmap-like datastructure. I *may* want to keep the cmap separate from the glyph data, as it helps preventing assigning multiple glyphs to a single code point

Reply:
The Component stuff is related to:
- https://github.com/BlackFoundryCom/variable-components-in-ufo/
- https://github.com/harfbuzz/boring-expansion-spec/blob/glyf1/glyf1.md

Reply:
I attempted to capture the key details from this thread in https://github.com/googlefonts/oxidize/commit/a3efa92f95b22e8bf1ee6d9a68e0d3b400111783. Closing, plz reopen or start a new thread if I missed key points, got it all backwards, etc.